### DISCUSSION TITLE
Custom loss function in spaCy 3

### DISCUSSION THREAD

**Discussion by davesuketu215** [explosion/spaCy#7295] [Help: Coding & Implementations] [answered]
As part of my requirement, I need to implement a customized loss function in my custom spaCy NER model while training.
Is there a provision in spaCy 3 to implement a custom scorer function (or loss function) in place of the standard ones?

We see that sklearn has make_scorer( ) which provides such provisions to use in the models. Is there any similar provision in spaCy? Are there any other ways in which this can be achieved in spaCy?

Any help would be appreciated

**Comment by svlandeg** [comment]
Hi! For a related question & answer, see here: https://github.com/explosion/spaCy/discussions/7256.

**Comment by svlandeg** [comment, marked as answer]
In short, there's not really a simple mechanism for this right now, no. The loss functions currently implemented in the built-in components are very much tailored towards their inner workings, so it might be a little involved to dive into this code, but it should still be doable to create a subclass that overwrites the loss calcalation.

**Reply by davesuketu215** [reply]
Thanks for the answer and related link. That helps a lot!! and I will experiment with it.

However, I have one confusion here - for NER model, I am unable to find the corresponding ```get_loss()``` function in ```cdef class EntityRecognizer(Parser):``` in the file ner.pyx

Is it somewhere else? Or am I searching at the wrong place? Please guide.

**Reply by davesuketu215** [reply]
Can somebody help me in finding out the correct place for the NER loss function? So that I can experiment with custom implementations of the same.

**Reply by adrianeboyd** [reply]
The transition-based parser `Parser` (for historical reasons?) has `get_batch_loss` instead of `get_loss`.

**Comment by mandar-avhad** [comment]
Hello, can you share how to compute accuracy of trained ner model  in spacy 3? Please share any source or code with reference to that. Thank you

**Reply by davesuketu215** [reply]
It seems there is already an API available from spaCy for the purpose:
Explore this: ```nlp.evaluate()```. You can even search the discussion forum more on this. spaCy documentation explains it here: https://spacy.io/api/language#evaluate
You need to convert your annotated validation set into Doc and then into Example objects and then pass on to this API.

Basically this will give you PRF Scores per entity type and also overall accuracy for NER and tokenizer modules. Basically it returns a nested dictionary data structure.

Hope this helps.

**Reply by mandar-avhad** [reply]
Yes, thank you.
But could you please tell how to convert our test data in examples format? How to add test data in nlp.evaluate?
